File: src/volumeController.ts
export type AppIdentifier = string;
export type VolumePercent = number & { __brand: "VolumePercent" }

export enum SessionType {
  Application = "Application",
  Device = "Device"
}

export enum SessionDirection {
  Render = "Render",
  Capture = "Capture",
  NOOP = "Noop"
}

export interface AudioSession {
  name: string;
  type: SessionType;
  direction: SessionDirection;
  deviceOutput: SessionDirection;
  deviceName: string;
  id: string;
  windowTitle: string;
  volumePercent: number;
  muted: boolean;
  active: boolean;
}

export abstract class VolumeController {
  abstract getPlaybackDevices(): Promise<AudioSession[]>;
  abstract getCurrentPlaybackDevice(): Promise<AudioSession | null>;
  abstract getMasterVolume(): Promise<number | null>;
  abstract setMasterVolume(percent: VolumePercent): Promise<void>;
  abstract muteMaster(): Promise<void>;
  abstract unmuteMaster(): Promise<void>;
  abstract setAppVolume(app: AppIdentifier, percent: VolumePercent): Promise<void>;
  abstract muteApp(app: AppIdentifier): Promise<void>;
  abstract unmuteApp(app: AppIdentifier): Promise<void>;
  abstract loadSessions(): Promise<AudioSession[]>;
  abstract getAllApplications(): Promise<AudioSession[]>;
}

File: src-tauri/src/VolumeController.rs
pub type VolumePercent = f32;
pub type AppIdentifier = String;

pub enum SessionType {
    Application = "Application",
    Device = "Device",
}

enum SessionDirection {
    Render = "Render",
    Capture = "Capture",
    NOOP = "Noop",
}

// #[derive(Debug, Clone)]
pub struct AudioSession {
    pub name: string,
    pub session_type: SessionType,
    pub direction: SessionDirection,
    pub device_output: SessionDirection,
    pub device_name: string,
    pub id: string,
    pub window_title: string,
    pub volume_percent: number,
    pub muted: boolean,
    pub active: boolean,
}

pub trait VolumeController {
    fn get_playback_devices(&self) -> Result<Vec<AudioSession>>;
    fn get_current_playback_device(&self) -> Result<Option<AudioSession>>;
    fn get_master_volume(&self) -> Result<Option<VolumePercent>>;
    fn set_master_volume(&self, percent: VolumePercent) -> Result<()>;
    fn mute_master(&self) -> Result<()>;
    fn unmute_master(&self) -> Result<()>;
    fn set_app_volume(&self, app: AppIdentifier, percent: VolumePercent) -> Result<()>;
    fn mute_app(&self, app: AppIdentifier) -> Result<()>;
    fn unmute_app(&self, app: AppIdentifier) -> Result<()>;
    fn load_sessions(&self) -> Result<Vec<AudioSession>>;
    fn get_all_applications(&self) -> Result<Vec<AudioSession>>;
}
